File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = 0
        for n in nums:
            total += n
        
        if total % 2 != 0:
            return False
        target = total >> 1

        dp = [False] * (target + 1)
        dp[0] = True

        for n in nums:
            for j in range(target, n - 1, -1):
                dp[j] = dp[j] or dp[j - n]
        
        return dp[target]

File: Data Structures & Algorithms/target-sum/submission-3.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number}
     */
    findTargetSumWays(nums, target) {
        const total = nums.reduce((a, b) => a + b, 0);
        target = Math.abs(target);
        const rem = total - target;
        if(rem < 0 || rem % 2 != 0) return 0;

        const k = rem >> 1;
        const dp = new Array(k + 1).fill(0);
        dp[0] = 1;

        for(const num of nums) {
            for(let i = k; i >= num; --i) {
                dp[i] += dp[i - num];
            }
        }

        return dp[k];
    }
}
